File: src/xyang/augment_expand.py
"""Resolve RFC 7950 ``augment`` targets and merge augmented schema into the target module.

Called from the parser only when ``expand_uses`` is True, so ``expand_uses=False`` keeps
``augment`` as explicit statements in the AST—matching the reversibility goal of
preserving ``uses``/grouping structure for round-trip convert paths.
"""

from __future__ import annotations

from typing import Callable, Dict, List, Optional

from .ast import (
    YangAugmentStmt,
    YangCaseStmt,
    YangChoiceStmt,
    YangStatement,
    YangStatementList,
)
from .module import YangModule
from .errors import YangSyntaxError
from .identifier_ref import YangIdentifierRef, parse_absolute_schema_path
from .refine_expand import copy_yang_statement
from .uses_expand import (
    _merge_uses_if_features_into_grouping_roots,
    _merge_uses_when_into_grouping_roots,
)


def _path_segments_from_augment(aug: YangAugmentStmt, path: str) -> List[YangIdentifierRef]:
    stored = aug.augment_path_segments
    if stored:
        return stored
    try:
        return parse_absolute_schema_path(path)
    except ValueError as exc:
        raise YangSyntaxError(str(exc)) from exc


def _find_named_schema_child(parent: YangStatement, name: str) -> Optional[YangStatement]:
    if isinstance(parent, YangChoiceStmt):
        for case in parent.cases:
            if case.name == name:
                return case
        return None
    if not hasattr(parent, "statements"):
        return None
    sl: YangStatementList = parent  # type: ignore[assignment]
    for s in sl.statements:
        if getattr(s, "name", None) == name and s.get_schema_node() is not None:
            return s
    return None


def _find_toplevel_schema_child(module: YangModule, name: str) -> Optional[YangStatement]:
    for s in module.statements:
        if getattr(s, "name", None) == name and s.get_schema_node() is not None:
            return s
    return None


def resolve_augment_target(
    ctx_module: YangModule, path: str, aug: YangAugmentStmt | None = None
) -> YangStatement:
    """
    Resolve an absolute augment path to the **target schema node** that receives new children.

    Each path segment is ``prefix:identifier`` (RFC 7950 absolute schema node identifier).
    """
    return resolve_absolute_schema_path(
        ctx_module=ctx_module,
        path=path,
        segments=_path_segments_from_augment(aug, path) if aug is not None else None,
        kind="augment",
        find_toplevel=_find_toplevel_schema_child,
    )


def resolve_absolute_schema_path(
    *,
    ctx_module: YangModule,
    path: str,
    kind: str,
    find_toplevel: Callable[[YangModule, str], Optional[YangStatement]],
    segments: List[YangIdentifierRef] | None = None,
) -> YangStatement:
    """Resolve absolute ``/prefix:name/...`` path into a target schema node.

    ``find_toplevel`` provides the root-node lookup strategy for segment 0.
    Child traversal semantics are shared (choice/case aware) via
    ``_find_named_schema_child``.
    """
    if not segments:
        try:
            segments = parse_absolute_schema_path(path)
        except ValueError as exc:
            raise YangSyntaxError(str(exc)) from exc
    first = segments[0]
    pref0 = first.prefix
    name0 = first.name
    if not pref0:
        raise YangSyntaxError(
            f"{kind}: invalid first segment in path {path!r} "
            "(expected 'prefix:identifier')"
        )
    mod0 = ctx_module.resolve_prefixed_module(pref0)
    if mod0 is None:
        raise YangSyntaxError(
            f"{kind}: unknown prefix {pref0!r} in path {path!r} "
            f"(module {ctx_module.name!r})"
        )
    cur = find_toplevel(mod0, name0)
    if cur is None:
        raise YangSyntaxError(
            f"{kind}: no top-level schema node {name0!r} in module {mod0.name!r} "
            f"(path {path!r})"
        )
    for seg in segments[1:]:
        pref = seg.prefix
        nm = seg.name
        if not pref:
            raise YangSyntaxError(
                f"{kind}: invalid path segment {nm!r} in path {path!r} "
                "(expected 'prefix:identifier')"
            )
        if ctx_module.resolve_prefixed_module(pref) is None:
            raise YangSyntaxError(
                f"{kind}: unknown prefix {pref!r} in path {path!r}"
            )
        nxt = _find_named_schema_child(cur, nm)
        if nxt is None:
            raise YangSyntaxError(
                f"{kind}: no child {nm!r} under node in path {path!r} "
                f"(after {cur.name!r})"
            )
        cur = nxt
    if not hasattr(cur, "statements"):
        raise YangSyntaxError(
            f"{kind}: target node {getattr(cur, 'name', '?')!r} cannot contain "
            f"schema substatements (path {path!r})"
        )
    return cur


def _stamp_defining_module(stmt: YangStatement, module_name: str) -> None:
    """Mark *stmt* and descendants as defined in *module_name* (RFC 7951 qualified JSON names)."""
    if hasattr(stmt, "defining_module"):
        stmt.defining_module = module_name  # type: ignore[attr-defined]
    for child in stmt.statements:
        _stamp_defining_module(child, module_name)
    if isinstance(stmt, YangChoiceStmt):
        for case in stmt.cases:
            _stamp_defining_module(case, module_name)


def _merge_augment_into_target(
    aug: YangAugmentStmt, target: YangStatement, *, source_module: YangModule
) -> None:
    """Copy augmented children onto *target* (``uses``-expanded body expected)."""
    copies = [copy_yang_statement(x) for x in aug.statements]
    for c in copies:
        _stamp_defining_module(c, source_module.name)
    _merge_uses_if_features_into_grouping_roots(copies, aug.if_features)
    _merge_uses_when_into_grouping_roots(copies, aug.when)
    if isinstance(target, YangChoiceStmt):
        for c in copies:
            if isinstance(c, YangCaseStmt):
                target.cases.append(c)
            else:
                target.statements.append(c)
    else:
        tlist: YangStatementList = target  # type: ignore[assignment]
        for c in copies:
            tlist.statements.append(c)


def register_module_closure(modules: Dict[str, YangModule], mod: YangModule) -> None:
    """Register *mod* and every module reachable via ``import`` (RFC 7950 import closure)."""
    modules[mod.name] = mod
    for imported in mod.import_prefixes.values():
        register_module_closure(modules, imported)


def apply_augmentations_across_module_map(modules: Dict[str, YangModule]) -> None:
    """
    Apply every top-level ``augment`` still present in any module in *modules*.

    Use after loading a set of related modules (e.g. anydata validation map) so
    augments defined in one file merge into targets in another module that shares
    the same :class:`YangModule` instances (one :class:`~xyang.parser.yang_parser.YangParser`
    cache per load batch).
    """
    pending: list[tuple[YangModule, YangAugmentStmt]] = []
    seen: set[int] = set()
    for mod in modules.values():
        oid = id(mod)
        if oid in seen:
            continue
        seen.add(oid)
        for stmt in mod.statements:
            if isinstance(stmt, YangAugmentStmt):
                pending.append((mod, stmt))
    for aug_module, aug in pending:
        target = resolve_augment_target(aug_module, aug.augment_path, aug)
        _merge_augment_into_target(aug, target, source_module=aug_module)
    for mod in modules.values():
        mod.statements = [
            s for s in mod.statements if not isinstance(s, YangAugmentStmt)
        ]


def apply_augmentations(root: YangModule) -> None:
    """
    For each top-level ``augment``, copy its (already ``uses``-expanded) children onto the
    target node, merge ``if-feature`` / ``when`` like ``uses``, then remove the
    ``augment`` statement from *root*.
    """
    augments = [s for s in root.statements if isinstance(s, YangAugmentStmt)]
    for aug in augments:
        target = resolve_augment_target(root, aug.augment_path, aug)
        _merge_augment_into_target(aug, target, source_module=root)
    root.statements = [s for s in root.statements if not isinstance(s, YangAugmentStmt)]

File: ts/src/transform/augment-expand.ts
/**
 * Resolve RFC 7950 ``augment`` targets and merge augmented schema into the target module.
 *
 * Called from the parser only when ``expandUses`` is True, so ``expandUses=false`` keeps
 * ``augment`` as explicit statements in the AST—matching the reversibility goal of
 * preserving ``uses``/grouping structure for round-trip convert paths.
 */

import { YangSyntaxError } from "../core/errors";
import { parseAbsoluteSchemaPath, type YangIdentifierRef } from "../core/identifier-ref";
import { SerializedStatement, YangModule } from "../core/model";

type ModuleData = Record<string, unknown>;

const SCHEMA_NODE_KEYWORDS = new Set([
  "container",
  "list",
  "leaf",
  "leaf-list",
  "choice",
  "case",
  "anydata",
  "anyxml",
  "notification",
  "rpc",
  "action",
  "input",
  "output"
]);

function deepCloneStatement(stmt: SerializedStatement): SerializedStatement {
  return JSON.parse(JSON.stringify(stmt)) as SerializedStatement;
}

function asStatements(value: unknown): SerializedStatement[] {
  return Array.isArray(value) ? (value as SerializedStatement[]) : [];
}

function ownPrefix(module: ModuleData): string {
  return String(module.prefix ?? "").replace(/^['"]|['"]$/g, "");
}

function resolvePrefixedModule(ctxModule: ModuleData, prefix: string): ModuleData | undefined {
  if (prefix === ownPrefix(ctxModule)) {
    return ctxModule;
  }
  const imports = ctxModule.import_prefixes as Record<string, ModuleData> | undefined;
  const resolved = imports?.[prefix];
  return resolved && typeof resolved === "object" ? resolved : undefined;
}

function pathSegmentsFromAugment(aug: SerializedStatement, path: string): YangIdentifierRef[] {
  const stored = aug.augment_path_segments;
  if (Array.isArray(stored) && stored.length > 0) {
    return stored as YangIdentifierRef[];
  }
  try {
    return parseAbsoluteSchemaPath(path);
  } catch (err) {
    throw new YangSyntaxError(err instanceof Error ? err.message : String(err));
  }
}

function isSchemaNode(stmt: SerializedStatement): boolean {
  const kw = typeof stmt.keyword === "string" ? stmt.keyword : "";
  return Boolean(stmt.name) && SCHEMA_NODE_KEYWORDS.has(kw);
}

function findNamedSchemaChild(parent: SerializedStatement, name: string): SerializedStatement | undefined {
  for (const child of asStatements(parent.statements)) {
    if (child.name === name && isSchemaNode(child)) {
      return child;
    }
  }
  return undefined;
}

function findToplevelSchemaChild(module: ModuleData, name: string): SerializedStatement | undefined {
  for (const child of asStatements(module.statements)) {
    if (child.name === name && isSchemaNode(child)) {
      return child;
    }
  }
  return undefined;
}

/**
 * Resolve an absolute augment path to the target schema node that receives new children.
 * Prefers parse-time ``augment_path_segments``; falls back to parsing ``path`` once.
 */
export function resolveAugmentTarget(
  ctxModule: ModuleData,
  path: string,
  aug?: SerializedStatement
): SerializedStatement {
  return resolveAbsoluteSchemaPath({
    ctxModule,
    path,
    segments: aug ? pathSegmentsFromAugment(aug, path) : undefined,
    kind: "augment",
    findToplevel: findToplevelSchemaChild
  });
}

export function resolveAbsoluteSchemaPath(options: {
  ctxModule: ModuleData;
  path: string;
  segments?: YangIdentifierRef[];
  kind: string;
  findToplevel: (module: ModuleData, name: string) => SerializedStatement | undefined;
}): SerializedStatement {
  const { ctxModule, path, kind, findToplevel } = options;
  let segments = options.segments;
  if (!segments || segments.length === 0) {
    try {
      segments = parseAbsoluteSchemaPath(path);
    } catch (err) {
      throw new YangSyntaxError(err instanceof Error ? err.message : String(err));
    }
  }
  const first = segments[0]!;
  const pref0 = first.prefix!;
  const name0 = first.name;
  const mod0 = resolvePrefixedModule(ctxModule, pref0);
  if (!mod0) {
    throw new YangSyntaxError(
      `${kind}: unknown prefix '${pref0}' in path '${path}' ` +
        `(module '${String(ctxModule.name ?? "")}')`
    );
  }
  let cur = findToplevel(mod0, name0);
  if (!cur) {
    throw new YangSyntaxError(
      `${kind}: no top-level schema node '${name0}' in module '${String(mod0.name ?? "")}' ` +
        `(path '${path}')`
    );
  }
  for (const seg of segments.slice(1)) {
    const pref = seg.prefix!;
    const nm = seg.name;
    if (!resolvePrefixedModule(ctxModule, pref)) {
      throw new YangSyntaxError(`${kind}: unknown prefix '${pref}' in path '${path}'`);
    }
    const nxt = findNamedSchemaChild(cur, nm);
    if (!nxt) {
      throw new YangSyntaxError(
        `${kind}: no child '${nm}' under node in path '${path}' ` + `(after '${String(cur.name ?? "")}')`
      );
    }
    cur = nxt;
  }
  if (!Array.isArray(cur.statements) && cur.statements !== undefined) {
    throw new YangSyntaxError(
      `${kind}: target node '${String(cur.name ?? "?")}' cannot contain ` +
        `schema substatements (path '${path}')`
    );
  }
  if (cur.statements === undefined) {
    cur.statements = [];
  }
  return cur;
}

function stampDefiningModule(stmt: SerializedStatement, moduleName: string): void {
  stmt.defining_module = moduleName;
  for (const child of asStatements(stmt.statements)) {
    stampDefiningModule(child, moduleName);
  }
}

function mergeIfFeaturesFromAugment(aug: SerializedStatement, child: SerializedStatement): void {
  const ufs = aug.if_features;
  if (!Array.isArray(ufs) || ufs.length === 0) {
    return;
  }
  const existing = Array.isArray(child.if_features) ? (child.if_features as string[]) : [];
  child.if_features = [...ufs, ...existing];
}

function mergeWhenFromAugment(aug: SerializedStatement, child: SerializedStatement): void {
  const uw = aug.when as Record<string, unknown> | undefined;
  if (!uw || typeof uw.expression !== "string" || uw.expression.trim() === "") {
    return;
  }
  const usesWhen: Record<string, unknown> = {
    ...deepCloneStatement(uw as SerializedStatement),
    evaluate_with_parent_context: true
  };
  const existing = child.when as Record<string, unknown> | undefined;
  if (!existing?.expression) {
    child.when = usesWhen;
    return;
  }
  child.when = {
    ...existing,
    expression: `(${String(existing.expression)}) and (${uw.expression})`,
    description: String(existing.description ?? ""),
    evaluate_with_parent_context: true
  };
}

function mergeAugmentIntoTarget(
  aug: SerializedStatement,
  target: SerializedStatement,
  sourceModuleName: string
): void {
  const copies = asStatements(aug.statements).map((x) => deepCloneStatement(x));
  for (const c of copies) {
    stampDefiningModule(c, sourceModuleName);
    mergeIfFeaturesFromAugment(aug, c);
    mergeWhenFromAugment(aug, c);
  }
  const targetChildren = asStatements(target.statements);
  for (const c of copies) {
    targetChildren.push(c);
  }
  target.statements = targetChildren;
}

/**
 * Register *mod* and every module reachable via ``import`` (RFC 7950 import closure).
 * ``import_prefixes`` values are module *data* records (shared with the parser cache).
 */
export function registerModuleClosure(modules: Map<string, YangModule>, mod: YangModule): void {
  const name = mod.name;
  if (!name) {
    return;
  }
  modules.set(name, mod);
  const imports = (mod.data.import_prefixes as Record<string, ModuleData> | undefined) ?? {};
  for (const importedData of Object.values(imports)) {
    if (!importedData || typeof importedData !== "object") {
      continue;
    }
    const importedName = typeof importedData.name === "string" ? importedData.name : "";
    if (!importedName || modules.has(importedName)) {
      continue;
    }
    // Wrap shared data so later merges mutate the same object the parser cached.
    registerModuleClosure(modules, new YangModule(importedData, { kind: "string", value: importedName }));
  }
}

/**
 * Apply every top-level ``augment`` still present in any module in *modules*.
 *
 * Use after loading a set of related modules so augments defined in one file merge into
 * targets in another module that shares the same module data instances (one
 * {@link YangParser} cache per load batch).
 */
export function applyAugmentationsAcrossModuleMap(modules: Map<string, YangModule> | Record<string, YangModule>): void {
  const list = modules instanceof Map ? [...modules.values()] : Object.values(modules);
  const pending: Array<{ mod: YangModule; aug: SerializedStatement }> = [];
  const seen = new Set<ModuleData>();
  for (const mod of list) {
    if (seen.has(mod.data)) {
      continue;
    }
    seen.add(mod.data);
    for (const stmt of asStatements(mod.data.statements)) {
      if (stmt.keyword === "augment") {
        pending.push({ mod, aug: stmt });
      }
    }
  }
  for (const { mod, aug } of pending) {
    const path = String(aug.augment_path ?? aug.argument ?? "");
    const target = resolveAugmentTarget(mod.data, path, aug);
    mergeAugmentIntoTarget(aug, target, String(mod.name ?? ""));
  }
  for (const mod of list) {
    const stmts = asStatements(mod.data.statements);
    mod.data.statements = stmts.filter((s) => s.keyword !== "augment");
  }
}

/**
 * For each top-level ``augment``, copy its (already ``uses``-expanded) children onto the
 * target node, merge ``if-feature`` / ``when`` like ``uses``, then remove the
 * ``augment`` statement from *module*.
 */
export function applyAugments(module: YangModule): YangModule {
  const data = module.data;
  const statements = asStatements(data.statements);
  const augments = statements.filter((s) => s.keyword === "augment");
  if (augments.length === 0) {
    return module;
  }
  const sourceName = String(data.name ?? "");
  for (const aug of augments) {
    const path = String(aug.augment_path ?? aug.argument ?? "");
    const target = resolveAugmentTarget(data, path, aug);
    mergeAugmentIntoTarget(aug, target, sourceName);
  }
  data.statements = statements.filter((s) => s.keyword !== "augment");
  return module;
}
